core/vm/interpreter.go

```diff
@@ -877,7 +878,7 @@ func (in *GethEVMInterpreter) Step(state *GethState) bool {
 		return false
 	}
 	// If the operation is valid, enforce write restrictions
-	if in.readOnly && in.evm.chainRules.IsByzantium {
+	if (in.readOnly || state.ReadOnly) && in.evm.chainRules.IsByzantium {
```

Comment:
I don't think this should be checked here. I would do this analogous to the way the `run` function does this. For `run` it gets the `readOnly` state as a function parameter, and then sets this in the `GethEVMInterpreter` struct.
Adding a public `ReadOnly` to the `GethState` is fine, but I think it should then be propagated to the private `GethEVMInterpreter.readOnly` just like `run` does, instead of manually checking both state representations here.

Reply:
Done